pkg/store/sorted_series_server.go

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package store
+
+import (
+	"sort"
+
+	"github.com/thanos-io/thanos/pkg/store/labelpb"
+	"github.com/thanos-io/thanos/pkg/store/storepb"
+)
+
+type sortedSeriesServer struct {
+	// This field just exist to pseudo-implement the unused methods of the interface.
+	storepb.Store_SeriesServer
+
+	responses           []*storepb.SeriesResponse
+	sortWithoutLabelSet map[string]struct{}
+	passThrough         bool
+}
+
+func newSortedSeriesServer(upstream storepb.Store_SeriesServer, sortWithoutLabelSet map[string]struct{}, passThrough bool) *sortedSeriesServer {
+	return &sortedSeriesServer{
+		Store_SeriesServer: upstream,
+
+		passThrough:         passThrough,
+		sortWithoutLabelSet: sortWithoutLabelSet,
+
+		// Buffered responses when passThrough is false.
+		responses: make([]*storepb.SeriesResponse, 0),
+	}
+}
+
+func (s *sortedSeriesServer) Send(r *storepb.SeriesResponse) error {
+	series := r.GetSeries()
+	// No need to buffer warnings and hints.
+	if series == nil {
+		return s.Store_SeriesServer.Send(r)
+	}
+
+	moveLabelsToEnd(series.Labels, s.sortWithoutLabelSet)
+	if s.passThrough {
+		return s.Store_SeriesServer.Send(r)
+	}
+
+	s.responses = append(s.responses, r)
+	return nil
+}
+
+func (s *sortedSeriesServer) Flush() error {
+	if s.passThrough {
+		return nil
+	}
+
+	if len(s.sortWithoutLabelSet) > 0 {
+		sort.Slice(s.responses, func(i, j int) bool {
+			return compareResponses(s.responses[i], s.responses[j])
+		})
+	}
+
+	for _, r := range s.responses {
+		if err := s.Store_SeriesServer.Send(r); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func moveLabelsToEnd(labelSet []labelpb.ZLabel, labelsToMove map[string]struct{}) {
+	if len(labelsToMove) == 0 {
+		return
+	}
+
+	sort.Slice(labelSet, func(i, j int) bool {
+		if _, ok := labelsToMove[labelSet[i].Name]; ok {
+			return false
+		}
+		if _, ok := labelsToMove[labelSet[j].Name]; ok {
+			return true
+		}
+
+		return labelSet[i].Name < labelSet[j].Name
+	})
+}
+
+func sortRequired(sortWithoutLabels map[string]struct{}, extLabelsMap map[string]struct{}) bool {
```

Comment:
For us this would mean forced buffering on StoreAPI nodes because we have replica labels set on Thanos Query that exist on some nodes and on some do not :/ thus, we'd be forced to set those external labels on all nodes, to some kind of fake values or something. I wonder how common is a setup like ours. :thinking:

Reply:
Actually we can solve this by looking at external labels on individual blocks. Will try to make this change.

Reply:
I updated store/bucket to sort only if replica labels are not found in at least one block. I think this should over the use case you described.

Reply:
Hm on second thought, the replica labels can be removed by compaction. In those cases, they won't be found in external labels and store-gateways will be forced to buffer. 

I don't know what would be a good solution for this case. We also might be asking for too much from the system without giving it the necessary information to make good decisions.

Reply:
Could we say that if _no_ label from `sortWithoutLabels` is present in `extLabelsMap` that sort is not required?

Reply:
Unfortunately that would be also true if a user wants to dedup by a label that is not external, and is present in series inside TSDB.

Reply:
Ah got it now in connection with your comment https://github.com/thanos-io/thanos/pull/5796#issuecomment-1288565538, I honestly also wasn't aware of this use case.